precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let relay_epoch_index =
+			<Runtime as pallet_randomness::Config>::BabeDataGetter::get_relay_epoch_index();
+		Ok(succeed(
+			EvmDataWriter::new().write(relay_epoch_index).build(),
+		))
+	}
+	/// Make request for babe randomness current block
+	fn request_babe_randomness_current_block(
+		handle: &mut impl PrecompileHandle,
+	) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let contract_address = handle.context().caller;
+		let refund_address = input.read::<Address>()?.0;
+		let fee: BalanceOf<Runtime> = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		let blocks_after_current = input.read::<u64>()?;
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let relay_block_number: u64 =
+			<Runtime as pallet_randomness::Config>::BabeDataGetter::get_relay_block_number()
+				.try_into()
+				.map_err(|_| revert("block number overflowed u64"))?;
+		let requested_block_number = blocks_after_current
+			.checked_add(relay_block_number)
+			.ok_or(error("addition result overflowed u64"))?
+			.try_into()
+			.map_err(|_| revert("u64 addition result overflowed block number type"))?;
+		let request = pallet_randomness::Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			salt: salt.into(),
+			info: pallet_randomness::RequestType::BabeCurrentBlock(requested_block_number),
+		};
+		pallet_randomness::Pallet::<Runtime>::request_randomness(request)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?;
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: Default::default(),
+		})
+	}
+	/// Make request for babe randomness one epoch ago
+	fn request_babe_randomness_one_epoch_ago(
+		handle: &mut impl PrecompileHandle,
+	) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let contract_address = handle.context().caller;
+		let refund_address = input.read::<Address>()?.0;
+		let fee: BalanceOf<Runtime> = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let next_epoch_index =
+			<Runtime as pallet_randomness::Config>::BabeDataGetter::get_relay_epoch_index()
+				.checked_add(1u64)
+				.ok_or(error("Epoch Index (u64) overflowed"))?;
+		let request = pallet_randomness::Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			salt: salt.into(),
+			info: pallet_randomness::RequestType::BabeOneEpochAgo(next_epoch_index),
+		};
+		pallet_randomness::Pallet::<Runtime>::request_randomness(request)
```

Comment:
You should use the new precompile logic using `RuntimeHelper::<Runtime>::try_dispatch(`, which records gas usage

Reply:
Is it not a dispatchable because we dont want it to be able to be called by signed origins? If so, why not require a special origin that we append in the precompile? Just in case it simplifies benchmarks

Reply:
> If so, why not require a special origin that we append in the precompile? Just in case it simplifies benchmarks

These benchmarks won't be that complicated so I don't think it's worth changing them to calls and gating them behind a special origin unless there was some other reason for doing this.

Reply:
The cost still need to be recorded. You should record some constant that will be updated once the benchmarks are done.